# Ex2 - Getting and Knowing your Data

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. Import the necessary libraries

In [1]:
import numpy as np
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

In [2]:
# assign data url to variable url
url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv"
pd.read_csv(url, delimiter='\t').head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

### Step 3. Assign it to a variable called chipo.

In [3]:
chipo = pd.read_csv(url, delimiter='\t')

### Step 4. See the first 10 entries

In [4]:
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

### Step 5. What is the number of observations in the dataset?

In [8]:
# Solution 1
print(f'There are {chipo.shape[0]} observations(rows) in the dataset.')


There are 4622 observations(rows) in the dataset.


In [9]:
# Solution 2
print(f'There are {len(chipo)} observations(rows) in the dataset.')



There are 4622 observations(rows) in the dataset.


### Step 6. What is the number of columns in the dataset?

In [10]:
print(f'There are {chipo.shape[1]} features(columns) in the dataset.')


There are 5 features(columns) in the dataset.


### Step 7. Print the name of all the columns.

In [30]:
print(f'{", ".join(chipo.columns)}')

order_id, quantity, item_name, choice_description, item_price


### Step 8. How is the dataset indexed?

In [33]:
chipo.index, chipo.head(1), chipo.tail(1)

(RangeIndex(start=0, stop=4622, step=1),
    order_id  quantity                     item_name choice_description  \
 0         1         1  Chips and Fresh Tomato Salsa                NaN   
 
   item_price  
 0     $2.39   ,
       order_id  quantity           item_name  \
 4621      1834         1  Chicken Salad Bowl   
 
                                      choice_description item_price  
 4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   )

Answer: Dataset has a Range index from 0 to 4621 by 1

### Step 9. Which was the most-ordered item? 

In [35]:
chipo.item_name.value_counts().head()

Chicken Bowl           726
Chicken Burrito        553
Chips and Guacamole    479
Steak Burrito          368
Canned Soft Drink      301
Name: item_name, dtype: int64

Answer: It looks like the Chicken Bowl is the most popular

### Step 10. For the most-ordered item, how many items were ordered?

726

### Step 11. What was the most ordered item in the choice_description column?

In [37]:
chipo.choice_description.value_counts().head()

[Diet Coke]                                                                          134
[Coke]                                                                               123
[Sprite]                                                                              77
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]                42
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Guacamole, Lettuce]]     40
Name: choice_description, dtype: int64

Diet Coke was the most ordered item in the choice_description column

### Step 12. How many items were orderd in total?

In [40]:
print(f'There were a total of {chipo.quantity.sum()} items ordered.')

There were a total of 4972 items ordered.


### Step 13. Turn the item price into a float

#### Step 13.a. Check the item price type

In [42]:
chipo.item_price.dtype

dtype('O')

Answer: The item price is a string(object).

#### Step 13.b. Create a lambda function and change the type of item price

In [47]:
chipo['item_price'] = chipo.item_price.apply(lambda x: float(x.replace('$', '').strip()))

#### Step 13.c. Check the item price type

In [48]:
chipo.item_price.dtype

dtype('float64')

### Step 14. How much was the revenue for the period in the dataset?

In [50]:
chipo[chipo.item_name == "Canned Soda"].head(10)

order_id  quantity    item_name choice_description  item_price
18          9         2  Canned Soda           [Sprite]        2.18
28         14         1  Canned Soda       [Dr. Pepper]        1.09
51         23         2  Canned Soda     [Mountain Dew]        2.18
53         24         1  Canned Soda           [Sprite]        1.09
107        47         1  Canned Soda       [Dr. Pepper]        1.09
117        51         1  Canned Soda  [Diet Dr. Pepper]        1.09
126        55         1  Canned Soda        [Coca Cola]        1.09
162        73         2  Canned Soda        [Diet Coke]        2.18
171        76         2  Canned Soda  [Diet Dr. Pepper]        2.18
179        81         1  Canned Soda        [Coca Cola]        1.09

In [54]:
# we can see here that the item_price reflects the quantity so revenue for the period
# should just be the sume of the item_price
print(f'The revenue for the period was ${chipo.item_price.sum():.2f}.')

The revenue for the period was $34500.16.


### Step 15. How many orders were made in the period?

In [57]:
        print(f'There were {len(chipo.order_id.unique())} orders made in this period.')

There were 1834 orders made in this period.


### Step 16. What is the average revenue amount per order?

In [59]:
# Solution 1
revenue = chipo.item_price.sum()
num_of_orders = len(chipo.order_id.unique())
print(f'The average revenue per order for the period was ${revenue / num_of_orders:.2f}.')


The average revenue per order for the period was $18.81.


In [65]:
# Solution 2
chipo.groupby('order_id').sum().mean()


quantity       2.711014
item_price    18.811429
dtype: float64

### Step 17. How many different items are sold?

In [66]:
chipo.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

In [67]:
len(chipo.item_name.unique())

50

In [111]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.7+ KB


In [112]:
# choice description is the only field with null values. These represent items
# where there is no sub-choice to be made e.g. chips(no flavors) v. soda(choice of flavors)
# So, let's replace the NaN values in the choice description with 'no_choice'
# so that the groupby can count it properly
chipo.choice_description.fillna('no_choice', inplace=True)

In [113]:
chipo.info()
# NaNs taken care of 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  4622 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.7+ KB


In [123]:
# now let's do a multi-layer groupby to see all of the item/subitem counts
item_groups = chipo.groupby(['item_name', 'choice_description']).count()
item_groups.head()
    

order_id  quantity  item_price
item_name         choice_description                                
6 Pack Soft Drink [Coke]                    21        21          21
                  [Diet Coke]               24        24          24
                  [Lemonade]                 2         2           2
                  [Nestea]                   2         2           2
                  [Sprite]                   5         5           5

In [131]:
# checking to make sure that we still have 50 items
len(item_groups.index.levels[0])

50

In [132]:
# now we can just check the length of choice description to get the total number
# of different items ordered this period
len(item_groups.index.levels[1])

1044